src/extensions/default/analyzers/Microsoft.DotNet.UpgradeAssistant.Extensions.Default.CSharp.Analyzers/ApiControllerAnalyzer.cs

```diff
@@ -0,0 +1,83 @@
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class ApiControllerAnalyzer : DiagnosticAnalyzer
+    {
+        public const string DiagnosticId = "UA0013";
+        private const string Category = "Upgrade";
+
+        public const string ApiControllerNamespace = "System.Web.Http";
+        public const string ApiControllerClassName = "ApiController";
+
+        private static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.ApiControllerDescription), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.ApiControllerDescription), Resources.ResourceManager, typeof(Resources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.ApiControllerDescription), Resources.ResourceManager, typeof(Resources));
+
+        protected static readonly DiagnosticDescriptor Rule = new(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterSymbolAction(AnalyzeSymbols, SymbolKind.NamedType);
+        }
+
+        private void AnalyzeSymbols(SymbolAnalysisContext context)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+            var baseType = namedTypeSymbol.BaseType;
+
+            if (baseType is null)
+            {
+                return;
+            }
+
+            if (IsBaseTypeAQualifiedReferenceToApiController(baseType)
+                || IsBaseTypeAnImplicitReferenceToApiController(baseType))
+            {
+                // For all such symbols, produce a diagnostic.
+                var node = namedTypeSymbol.DeclaringSyntaxReferences[0].GetSyntax();
+                ReportDiagnostic(context, node);
+            }
+        }
+
+        private static bool IsBaseTypeAnImplicitReferenceToApiController(INamedTypeSymbol baseType)
+        {
+            return baseType.ToDisplayString().Equals(ApiControllerClassName, StringComparison.Ordinal) && baseType.TypeKind == TypeKind.Error;
+        }
+
+        private static bool IsBaseTypeAQualifiedReferenceToApiController(INamedTypeSymbol baseType)
+        {
+            return baseType.ToDisplayString().Equals($"{ApiControllerNamespace}.{ApiControllerClassName}", StringComparison.Ordinal);
```

Comment:
you might want to consider making constants out of these concatenated strings.  otherwise you'll generate this for every type you look at.  allocations can affect perf.  but as you may be a run-once set of analyzers, that might be ok.